src/Table.hxx

```diff
@@ -306,9 +379,74 @@ public:
 
     void shrink_ipac_string_columns_to_fit(const std::vector<size_t> &column_widths);
 
+    std::string extract_value_as_string(const std::string &col_name,
+                                        size_t row_id) const;
+    std::string extract_value_as_string(size_t col_id, size_t row_id) const;
+
     std::vector<std::string> extract_column_values_as_strings(
             const std::string &colname) const;
 
+    template <typename T>
+    std::vector<T> extract_value(const std::string &col_name, size_t row_id) {
+        auto col_id = column_index(col_name);
+        return extract_value<T>(col_id, row_id);
+    }
+
+    template <typename T>
+    std::vector<T> extract_value(size_t col_id, size_t row_id) {
+        if (col_id >= columns.size()) {
+            throw std::runtime_error("Invalid column index: " + std::to_string(col_id));
+        }
+        if (row_id >= num_rows()) {
+            throw std::runtime_error("Invalid row index: " + std::to_string(row_id));
+        }
+
+        std::vector<T> val_array;
+        auto &column = columns[col_id];
+        auto array_size = column.array_size;
+        size_t row_offset = row_id * row_size();
+        if (is_null(row_offset, col_id)) {
+            for (size_t i = 0; i < array_size; ++i) {
+                val_array.emplace_back(get_null<T>());  // JTODO
+            }
+        } else {
+            // JTODO what if an element is null?  Assume already has get_null() value?
+            // JTODO Error (here and elsewhere) if not fixed array_size?
+            size_t base_offset = row_offset + offsets[col_id];
+            uint8_t const *curr_data = data.data() + base_offset;
+            size_t element_size = data_size(column.type);
+
+            for (size_t i = 0; i < array_size; ++i) {
+                val_array.emplace_back(*(reinterpret_cast<const T *>(curr_data)));
+                curr_data += element_size;
+            }
+        }
+        return val_array;
+    }
+
+
+    template <typename T>
+    std::vector<std::vector<T>> extract_column(const std::string &col_name) {
+        auto col_id = column_index(col_name);
+        return extract_column<T>(col_id);
+    }
+
+    template <typename T>
+    std::vector<std::vector<T>> extract_column(size_t col_id) {
+        if (col_id >= columns.size()) {
+            throw std::runtime_error("Invalid column index: " + std::to_string(col_id));
+        }
+
+        std::vector<std::vector<T>> col_vec;
+        auto &column = columns[col_id];
+
+        for (size_t curr_row_id = 0; curr_row_id < num_rows(); ++curr_row_id) {
+            col_vec.emplace_back(extract_value<T>(col_id, curr_row_id));
+        }
+        return col_vec;
+    }
```

Comment:
Re the column of scalars, would you prefer an extract_scalar_column() function that assumes array_size == 1 and returns std::vector<T> ? 
Re strings: right, this footprint doesn't look appropriate anyway for columns of type char.  Maybe an extract_char_column() that basically does extract_column_values_as_strings() for char columns?

Reply:
Re the column of scalars, would you prefer an extract_scalar_column() function that assumes array_size == 1 and returns std::vector<T> ? 
Re strings: right, this footprint doesn't look appropriate anyway for columns of type char.  Maybe an extract_char_column() that basically does extract_column_values_as_strings() for char columns?

Reply:
I was thinking you could return a plain vector and let clients deal with the fact that each row corresponds to n elements, but maybe that’s just too ugly of an API. For char columns, I think your suggestion returning a vector of strings sounds good.

Reply:
What would the type be of a plain vector in which each row corresponds to n uint (say) values?

Reply:
We already have extract_column_values_as_strings()  if the caller is willing to get values in string form.

Reply:
std::vector<uint32_t> - the consumer would have to index into it via row_index*n

Reply:
Of course this only works for fixed array sizes, but that’s all this library supports, right?

Reply:
Re: extract_column_values_as_strings - I missed that, thanks for pointing it out! I don’t think anything else is needed then.

Reply:
Ah. I'll check. There's a comment in write_hdf5_attributes.cxx that mentions variable-length arrays but I don't know the context:

>    /// This gets a bit complicated because there are variable length
>   /// arrays inside variable length arrays.

Reply:
Angela did ask for a way to extract a vector of the appropriate type.  I'm fine with your suggestion of a plain vector with n elements per row.

Reply:
Er.  Modulo the variable-length arrays question.

Reply:
But really, everything about this code assumes that arrays have fixed length.